### PR TITLE
Length validators: Support all unicode characters

### PR DESCRIPTION
**Pull request by DybekK**
Follow up to https://github.com/softwaremill/tapir/issues/3201. This PR adds optional support for counting unicode characters during validation.

### CODE REVIEW

**Comment by adamw** [comment]
There are some compilation failures. Some user-facing scaladoc on what the parameter means would also be nice.

Can you maybe also share what the specs (RFC) say about emojis as query/header/path values? Are they allows, should they be encoded?

File: core/src/main/scala/sttp/tapir/Validator.scala [hunk 1]
```diff
@@ -111,11 +112,17 @@ object Validator extends ValidatorMacros {
   case class Pattern[T <: String](value: String) extends Primitive[T] {
     override def doValidate(t: T): ValidationResult = ValidationResult.validWhen(t.matches(value))
   }
-  case class MinLength[T <: String](value: Int) extends Primitive[T] {
-    override def doValidate(t: T): ValidationResult = ValidationResult.validWhen(t.size >= value)
+  case class MinLength[T <: String](value: Int, countCodePoints: Boolean = false) extends Primitive[T] {
```
Comment (kciesielski) [inline core/src/main/scala/sttp/tapir/Validator.scala:115]: **suggestion:** A scaladoc explaining the purpose `countCodePoints` may be useful here.

**Review by kciesielski — COMMENTED** [review]
How about naming the PR "Length validators: Support all unicode characters", because "Add support for unicode characters" doesn't really say where exactly we're adding this support, and it suggest that unicode wasn't supported at all.

**Comment by DybekK** [comment]
The [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) specifies that URIs should consist of a limited set of characters: basic Latin alphabet letters, digits, and a few special characters. It doesn't explicitly mention emojis, but according to [Section 2.1 of RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-2.1) any character outside the allowed set should be percent-encoded, meaning that emojis can be a part of URI syntax when they're properly encoded.

**Comment by kciesielski** [comment]
Emojis can be encoded, but then we get for example `%F0%9F%98%8D` to represent "😍" in an URI, so length of such text should equal `12`, because for URI we treat it as length in ascii characters, right?

**Comment by adamw** [comment]
@DybekK thanks, I would put that in the docs - that the only scenario where this should be used is for body validators, where one may wish to count code points instead of characters (maybe giving an example of emojis ;) )

**Comment by adamw** [comment]
@kciesielski ah good point, we validate decoded strings. So this is also useful for queries etc., just depends on what you want to count.

File: core/src/main/scala/sttp/tapir/Validator.scala [hunk 2]
```diff
@@ -46,9 +46,43 @@ object Validator extends ValidatorMacros {
...
+  def minLength[T <: String](value: Int, countCodePoints: Boolean = false): Validator.Primitive[T] = MinLength(value, countCodePoints)
+
+  /** Create a validator for maximum length constraints on strings.
+    *
+    * @param value
+    *   The maximum allowed length.
+    * @param countCodePoints
+    *   A boolean parameter that determines whether the validation will consider code points or character count. When set to true, the
+    *   validator will consider characters that are represented using two code units (a surrogate pair) in UTF-16 encoding allowing for a
+    *   more accurate length validation for strings containing such characters. Defaults to false, where length is validated based on the
```
Comment (kciesielski) [inline core/src/main/scala/sttp/tapir/Validator.scala:69]: **nitpick:** "such characters" -> "such characters (like emojis)", maybe such a hint will additionally help people to understand if this is their case.

**Review by kciesielski — APPROVED** [review]

**Comment by adamw** [comment]
Thanks!

**Pull request merged**